File: Java/src/main/java/com/example/sort/MergeSort.java
package com.example.sort;
// 归并排序
public class MergeSort {

    static public void sort(int[] arr, int[] temp, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            sort(arr, temp, left, mid);
            sort(arr, temp, mid + 1, right);
            merge(arr, temp, left, mid, right);
        }
    }

    static public void merge(int[] arr, int[] temp, int left, int mid, int right) {
        int i = left;
        int j = mid + 1;
        int k = left;
        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }
        while (i <= mid) {
            temp[k++] = arr[i++];
        }
        while (j <= right) {
            temp[k++] = arr[j++];
        }

        for (i = left; i <= right; i++) {
            arr[i] = temp[i];
        }
    }

}

File: typescript/sort/mergeSort.ts
// 归并排序
export function mergeSort(arr: number[], left: number = 0, right: number = arr.length - 1, temp: number[] = []) {
    if (left >= right) return
    const middle = (left + right) >> 1
    mergeSort(arr, left, middle, temp)
    mergeSort(arr, middle + 1, right, temp)
    merge(arr, left, middle, right, temp)
}

export function merge(arr: number[], left: number, mid: number, right: number, temp: number[]) {
    let i = left, j = mid + 1, k = left;
    // 利用 temp 将两个有序数组合并
    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++]
        } else {
            temp[k++] = arr[j++]
        }
    }
    // 两个子数组可能不一样长
    while (i <= mid) {
        temp[k++] = arr[i++]
    }
    while (j <= right) {
        temp[k++] = arr[j++]

    }
    // 将 temp 中的元素拷贝回原数组
    for (i = left; i <= right; i++) {
        arr[i] = temp[i]
    }
}